### PR TITLE
initial commit for ui-ux section

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
This PR adds the initial UI/UX section. Going to leave is WIP for now but I do encourage folks to start commenting on it if they wish. 

## For all pull requests

- [ ] Link to the issue you are solving (so it gets closed)
- [x] Label the pull request with the appropriate area(s)
- [x] Assign potential reviewers (you may also want to contact them on Teams to ensure timely reviews)
- [x] Assign the project - Engineering Playbook Backlog
- [x] Assign the pull request to yourself

### CODE REVIEW

File: docs/ui-ux/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+   Story: I start by accessing the application in my browser and with my audio-aide enabled. I am immediately focussed on the authentication form and enter my corporate information. Once I land on the main page, I'm focussed on the search box. I tab to the image list and use my aide to listen to the first 10 images displayed. I tab back to the search box and type "dog". I then tab to the image list against and listen to the list of images, now all returning some form of my search "dog".
+
+User stories help you and your team understand application and feature requirements. In application design, the user drives the features. By completely understanding your user, you in turn, can completely understand what you need to build.
+
+Following the user stories, the design document should detail the accessibility requirements of the application. **Accessibility is never optional**. Microsoft has made a public commitment to deliver No matter who you determine your users are, there is a strict set of minimum requirements for all [Microsoft applications](https://www.microsoft.com/en-us/accessibility).
```
Comment (KatieProchilo) [inline docs/ui-ux/README.md:37]: Looks like you're either missing a period or maybe meant to remove something from the last sentence

File: docs/ui-ux/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+1. Persona: A user that can see and use a mouse without assistance.
+   Story: I start by accessing the application in my browser. I log in using my corporate account, and then see a search box and a list of images. I start typing "dog" into the search box. The images begin shuffling around, with ones tagged "dog" appearing at the top.
+1. Persona: A user that requires an audio-aide for browsing the web since they cannot see.
+   Story: I start by accessing the application in my browser and with my audio-aide enabled. I am immediately focussed on the authentication form and enter my corporate information. Once I land on the main page, I'm focussed on the search box. I tab to the image list and use my aide to listen to the first 10 images displayed. I tab back to the search box and type "dog". I then tab to the image list against and listen to the list of images, now all returning some form of my search "dog".
+
+User stories help you and your team understand application and feature requirements. In application design, the user drives the features. By completely understanding your user, you in turn, can completely understand what you need to build.
```
Comment (KatieProchilo) [inline docs/ui-ux/README.md:35]: Remove last comma?

File: docs/ui-ux/README.md (hunk 1 above)
Comment (KatieProchilo) [reply docs/ui-ux/README.md:37]: I couldn't find the "strict set of minimum requirements" by navigating to that link. Would these only come up once you download the browser tool below?

File: docs/ui-ux/README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+After listing accessibility requirements, the last section of the design document is technical considerations. This section outlines any and all implementation options your team wants to consider. Within CSE, we often reach for tools such as the [React] framework. React is a great tool when wielded by an experienced team. Otherwise, it can create more hurdles than it is worth. Keep in mind that even if _you_ feel capable with React, the rest of your team and your customer's dev team needs to as well. Some other great options to consider when building a proof-of-concept or minimum-valuable-product are:
+
+1. HTML/CSS/JavaScript
+   - Back to the basics! Start with a single **index.html**, include a popular CSS framework such as [Bootstrap]() using their CDN link, and start prototyping!
+   - Rarely will you have to support legacy browsers; thus, you can rely on modern JavaScript language features! No need for build tools or even TypeScript (did you know you can [type check JavaScript]()?)
```
Comment (KatieProchilo) [inline docs/ui-ux/README.md:49]: Missing links here, maybe part of the WIP 🙂

File: docs/ui-ux/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+1. Web Component frameworks
+   - Web Components are now standardized in all modern browsers
+   - Microsoft has their own, stable & actively-maintained framework, [Fast](https://fast.design)
```
Comment (KatieProchilo) [inline docs/ui-ux/README.md:52]: Not sure that my feedback will belong here since I love the casual/flexible/open-ended language you've used and this is just a README but...

I'd find it hugely helpful to point out the signs that maybe React is not the best choice and it's time for HTML instead or Web Components. Again, that seems a bit technical for the engineering playbook but would be interested to learn

File: docs/ui-ux/README.md (hunk 1 above)
Comment (Ethan-Arrowood) [reply docs/ui-ux/README.md:37]: good catch. I rewrote this paragraph so it made more sense. You're right, there isn't exactly a "strict set of minimum requirements". But like every app needs to pass the accessibility insights tool. I'm going to go into much more detail on this in the accessibility page.

File: docs/ui-ux/README.md (hunk 4 above)
Comment (Ethan-Arrowood) [reply docs/ui-ux/README.md:52]: Yeah i'm still pondering how to best author this topic. One comment I'm really happy with is in the beginning of this page: 

> Keep in mind that like all software, there is no "right way" to build a user interface application. Leverage and trust your team's or your customer's experience and expertise for the best development experience.

I feel like at the end of the day I want CSE crews to use what they are most comfortable with. What I don't want crews to do is to pick React because its what everyone else is using (insert a joke about friends and bridges here).

File: docs/ui-ux/README.md [hunk 5]
```diff
@@ -1,16 +1,80 @@
...
+
+User stories help you and your team understand application and feature requirements. In application design, the user drives the features. By completely understanding your user, you in turn, can completely understand what you need to build.
+
+Following the user stories, the design document should detail the accessibility requirements of the application. **Accessibility is never optional**. Microsoft has made a public commitment to deliver No matter who you determine your users are, there is a strict set of minimum requirements for all [Microsoft applications](https://www.microsoft.com/en-us/accessibility).
+
+> If you haven't already, download the [Accessibility Insights](https://accessibilityinsights.io/) browser tool. You can use this, in combination with other accessibility-related tools, to easily check your application.
```
Comment (engineersamuel) [inline docs/ui-ux/README.md:39]: Also I believe there is a node module or similar app which one can put as part of the CI/CD process to report out accessibility issues and how accessible a site is every time it is built.  Maybe look that up and include it here as well for automation.

File: docs/ui-ux/README.md (hunk 3 above)
Comment (engineersamuel) [reply docs/ui-ux/README.md:49]: I may suggest rewording this point, for example:

- Typically, the default output for client-side JavaScript is sufficient for an MVP app, if running in production and targetting legacy browsers (which happens but is more uncommon) then look into polyfills and configuration such as (browserlist)[https://github.com/browserslist/browserslist]. 
- Typescript is recommended for new projects however JavaScript can be sufficient as well, but it lacks the robust Typing system of Typescript.

File: docs/ui-ux/README.md (hunk 5 above)
Comment (Ethan-Arrowood) [reply docs/ui-ux/README.md]: That will be good for the accessibility document

**Comment by Ethan-Arrowood** [comment]
Publishing this PR as is so we can get it merged asap and then iterate on the incomplete sections. Hopefully after the M+C all hands I'll have more folks interested in authoring content too. @engineersamuel @KatieProchilo if y'all want to give this another look that would be appreciated. if you have additional content ideas maybe we can put them into an issue and include them in a follow up PR soon? I'd like to have a page to point to before Wednesday

File: docs/ui-ux/README.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
+# User Interface and User Experience Engineering
+
+Also known as _UI/UX_, _Front End Development_, or _Web Development_, user interface and user experience engineering is a broad topic and encompasses many different aspects of modern application development. When a user interface is required, CSE primarily develops a **web application**. Web apps can be built in a variety of ways with many different tools.
+
+## Goal
+
+The goal of the **User Interface Engineering Fundamentals** group is to provide guidance on developing web applications for CSE engagements. Everyone should begin by reading the [General Guidance](#general-guidance) for a quick introduction to the four main aspects of every web application project. From there, readers are encouraged to dive deeper into each topic or begin reviewing technical guidance that pertains to their engagement. All UI/UX projects should begin with a detailed design document. Review the [Design Document](#design-document) section for more details and a template to get your crew started.
```
Comment (TessFerrandez) [inline docs/ui-ux/README.md:7]:
```suggestion
The goal of the **User Interface** section is to provide guidance on developing web applications for CSE engagements. Everyone should begin by reading the [General Guidance](#general-guidance) for a quick introduction to the four main aspects of every web application project. From there, readers are encouraged to dive deeper into each topic, or begin reviewing technical guidance that pertains to their engagement. All UI/UX projects should begin with a detailed design document. Review the [Design Document](#design-document) section for more details, and a template to get started.
```

We should see the playbook as standalone from the EF groups - and ideally shouldn't reference internal groups or processes. The playbook should ideally be readable and applicable also to teams outside of CSE.

File: docs/ui-ux/README.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+Keep in mind that like all software, there is no "right way" to build a user interface application. Leverage and trust your team's or your customer's experience and expertise for the best development experience.
+
+## General Guidance
+
+The state of web platform engineering is fast moving. There is no one-size-fits-all solution. For any dev crew to be successful in building a UI, they need to have an understanding of the higher-level aspects of all UI project.
```
Comment (TessFerrandez) [inline docs/ui-ux/README.md:13]:
```suggestion
The state of web platform engineering is fast moving. There is no one-size-fits-all solution. For any team to be successful in building a UI, they need to have an understanding of the higher-level aspects of all UI project.
```

File: docs/ui-ux/README.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+1. [**Stability**](./stability.md) - what is the cost of adding a dependency? Is it actively stable/updated/maintained? If not, can you afford the tech debt (sometimes the answer can be yes!)? Could you get 90% of the way there without adding another dependency?
+
+More information is available for each general guidance section in the corresponding pages.
+
+## Design Document
+
+When beginning a UI/UX project, the first step is to produce a detailed design document that outlines the prospective solution. Design documents are meant to be a starting point, and unlike a gameplan document, they can change throughout an engagement.
+
+> Be certain to not over-promise on UI/UX requirements in the gameplan document. CSE teams should **not** agree to delivering production-ready user interface applications. The engineering complexities and requirements are far too much for an engagement. Committing to proof-of-concept or minimum-valuable-product applications is significantly more achievable within the timeline of a single (or even multiple) engagements.
```
Comment (TessFerrandez) [inline docs/ui-ux/README.md:26]: This refers to the internal game plan - can we rephrase this to make it more general?

File: docs/ui-ux/README.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+1. HTML/CSS/JavaScript
+   - Back to the basics! Start with a single **index.html**, include a popular CSS framework such as [Bootstrap](https://getbootstrap.com/) using their CDN link, and start prototyping!
+   - Rarely will you have to support legacy browsers; thus, you can rely on modern JavaScript language features! No need for build tools or even TypeScript (did you know you can [type check JavaScript](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html)).
+1. Web Component frameworks
+   - Web Components are now standardized in all modern browsers
+   - Microsoft has their own, stable & actively-maintained framework, [Fast](https://fast.design)
+
+For more information of choosing the right implementation tool, read the [Recommended Technologies](./recommended-technologies.md) document.
+
+### Template
```
Comment (TessFerrandez) [inline docs/ui-ux/README.md:61]: Should this template be split out to a separate file, so teams can download and use it in their projects?

File: docs/ui-ux/recommended-technologies.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+> TypeScript is JavaScript with syntax for types. TypeScript is a strongly typed programming language that builds on JavaScript, giving you better tooling at any scale.
+> [typescriptlang.org](https://www.typescriptlang.org/)
+
+TypeScript is highly recommended for all new web application projects. The stability it provides for teams is unmatched, and can make it easier for folks with C# backgrounds to work with web technologies.
+
+There are many ways to integrate TypeScript into a web application. The easiest way to get started is by reviewing the [TypeScript Tooling in 5 Minutes](https://www.typescriptlang.org/docs/handbook/typescript-tooling-in-5-minutes.html) guide from the official TypeScript docs. The other sections on this page contain information regarding integration with TypeScript.
+
+## React
+
+Easily the most commonly selected framework for CSE UI projects, React is a framework developed and maintained by Facebook. React is used throughout Microsoft and has a vast open source community.
```
Comment (TessFerrandez) [inline docs/ui-ux/recommended-technologies.md:20]:
```suggestion
React is a framework developed and maintained by Facebook. React is used throughout Microsoft and has a vast open source community.
```

**Review by TessFerrandez — APPROVED** [review]
Overall great start... I have approved the PR to not delay, but would like to see references to internal processes removed before the content is merged.

**Comment by Ethan-Arrowood** [comment]
Thank you for the great review @TessFerrandez - i've made all your recommended changes 😁

File: docs/ui-ux/README.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+1. [**Maintainability**](./maintainability.md) - is the application just a proof of concept to showcase an idea for future work, or will it be an MVP and act as the starting point for a larger, production-ready application? Sometimes you don't need React or any other framework. Sometimes you need React, but not all the bells and whistles from create-react-app. Understanding project maintainability requirements can simplify an engagement’s tooling needs significantly and let folks iterate without headaches.
+1. [**Stability**](./stability.md) - what is the cost of adding a dependency? Is it actively stable/updated/maintained? If not, can you afford the tech debt (sometimes the answer can be yes!)? Could you get 90% of the way there without adding another dependency?
+
+More information is available for each general guidance section in the corresponding pages.
+
+## Design Document
+
+When beginning a UI/UX project, the first step is to produce a detailed design document that outlines the prospective solution. Design documents are meant to be a starting point, and can change throughout an engagement.
```
Comment (wbreza) [inline docs/ui-ux/README.md:24]: It would be great to cross link into some of the "design" area document and work within them on any required revisions to the design docs area.

File: docs/ui-ux/README.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+   - Rarely will you have to support legacy browsers; thus, you can rely on modern JavaScript language features! No need for build tools or even TypeScript (did you know you can [type check JavaScript](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html)).
+1. Web Component frameworks
+   - Web Components are now standardized in all modern browsers
+   - Microsoft has their own, stable & actively-maintained framework, [Fast](https://fast.design)
+
+For more information of choosing the right implementation tool, read the [Recommended Technologies](./recommended-technologies.md) document.
+
+### Template
+
+This template is also available as a standalone file available here: [design-document-template.md](./design-document-template.md)
```
Comment (wbreza) [inline docs/ui-ux/README.md:63]: Please have a discussion with @zmmille2 around merging int design related content if applicable.

**Review by wbreza — APPROVED** [review]
Great start.  Added some feedback around incorporating design requirements into the design EF area as neded.

File: docs/ui-ux/recommended-technologies.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
+# Recommended Technologies
+
+The purpose of this page is review the commonly selected technology options when developing user interface applications.The reiterate from the general guidance section:
```
Comment (luisamiranda) [inline docs/ui-ux/recommended-technologies.md:3]: typo: "The purpose of this page is to review..."
typo: "To reiterate..."

File: docs/ui-ux/README.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+Following the user stories, the design document should detail the accessibility requirements of the application. **Accessibility is never optional**. No matter who you determine your users are, Microsoft has made a public commitment to always produce accessible appliations. For more information visit the official [Microsoft accessibility site](https://www.microsoft.com/en-us/accessibility).
+
+> If you haven't already, download the [Accessibility Insights](https://accessibilityinsights.io/) browser tool. You can use this, in combination with other accessibility-related tools, to easily check your application for accessibility issues.
+
+Accessibility requirements vary from project-to-project based on your intended user base. Aside from the default level of accessibility, this section should contain details about any other aspects you must consider during development. Could the application be used by someone who cannot see? What about someone with color blindness? What about someone who cannot use a mouse? If your application has an audio aspect, what about users that cannot hear?
```
Comment (luisamiranda) [inline docs/ui-ux/README.md:46]: typo: "project to project" does not need hyphens.

**Review by luisamiranda — APPROVED** [review]
This is going to be really useful. Love the focus on accessibility and the template will be handy to standardize practices in CSE. I look forward to seeing future sections. I just pointed out a few typos.

File: docs/ui-ux/README.md (hunk 11 above)
Comment (Ethan-Arrowood) [reply docs/ui-ux/README.md:24]: Great point. I'll make note of this and will make the modifications in a follow up. Thank you!

File: docs/ui-ux/README.md (hunk 12 above)
Comment (Ethan-Arrowood) [reply docs/ui-ux/README.md:63]: Same as previous. I've also reached out to Zach to discuss.

**Comment by Ethan-Arrowood** [comment]
Thank you everyone who reviewed ❤️. I'm going to merge this once the CI passes. Other reviewers are welcome to continue commenting on this PR; I will be going back through tomorrow and opening issues for all of the work that needs to be done.
